library/src/main/java/com/pokegoapi/api/PokemonGo.java

```diff
@@ -192,14 +203,40 @@ public void login(CredentialProvider credentialProvider)
 		playerProfile = new PlayerProfile(this);
 		playerProfile.updateProfile();
 
+		try {
+			if (hasChallenge()) {
+				Log.d(TAG, "Challenge active!");
+				awaitChallenge();
+			}
+		} catch (InterruptedException e) {
+			throw new LoginFailedException(e);
+		}
+
 		initialize();
 
 		this.loggingIn = false;
+
+		heartbeat.start();
 	}
 
 	private void initialize() throws RemoteServerException, CaptchaActiveException, LoginFailedException {
-		fireRequestBlock(new ServerRequest(RequestType.DOWNLOAD_REMOTE_CONFIG_VERSION,
-				CommonRequests.getDownloadRemoteConfigVersionMessageRequest()));
+		ServerRequest request = new ServerRequest(RequestType.DOWNLOAD_REMOTE_CONFIG_VERSION,
```

Comment:
I dont remember all of them on top of my head but there ASSET_DIGEST, PLAYER_PROFILE, LEVEL_UP_REWARDS and some others